Delegate implementation for Valve Configuration and control cluster. Tested with watervalve CHEF device.

sxb427 opened the pull request:
Adds a delegate implementation for valve configuration and control cluster intended to be used by chef devices using this cluster. Summary of changes -

- External Read / Write callbacks for `RemainingTime` attribute which has only external storage option available.
- Command handlers : 
    - HandleOpenValve: Returns the same level passed as input.
    - HandleCloseValve: NOP for now. TBD if `currentState` should be set to `kClosed` in server or delegate.
    - HandleRemainingDurationTick: NOP.
- Attribute modifications for watervalve device (ZAP):
    - DefaultOpenDuration default value to 100: Valve will remain in kOpen state after Open command is sent, for DefaultOpenDuration seconds.
    - Featuremap default value to 2: Enable `kLevel` feature for device.

#### Testing with watervalve

```
# Commission
./chip-tool pairing code 0x20 MT:Y.K90AFN00KA0648G00

# Read Current state (should be 0: kClosed)
./chip-tool valveconfigurationandcontrol read current-state 0x20 1 | grep "CurrentState"

# Send Open command with level = 50%. This will start a timer from 100 and continuously update RemainingDuration.
./chip-tool valveconfigurationandcontrol command-by-id 0 '{"0x1": 50}' 0x20 1
./chip-tool valveconfigurationandcontrol read remaining-duration 0x20 1 | grep "RemainingDuration"

# CurrentLevel should be set to 50 and, currentState to kOpen: 1
./chip-tool valveconfigurationandcontrol read current-level 0x20 1 | grep "CurrentLevel"
./chip-tool valveconfigurationandcontrol read current-state 0x20 1 | grep "CurrentState"

# Send close command. CurrentState changes to 2: kTransitioning
./chip-tool valveconfigurationandcontrol command-by-id 1 '{}' 0x20 1
./chip-tool valveconfigurationandcontrol read current-state 0x20 1 | grep "CurrentState"
```

sxb427 committed:
Enable new attributes on valve-configuration-and-control-cluster for watervalve

sxb427 committed:
Add delegate for valve-configuration-and-control cluster

sxb427 committed:
Fix remaining bugs - delegate not initialized, null ptr referrence in handleOpenValve, kLevel (0x2) has to be set in feature bitmap for Open command to set the right currentLevel.

sxb427 committed:
Fix Ifdefs and includes

andy31415 commented on src/app/clusters/valve-configuration-and-control-server/valve-configuration-and-control-server.h:
```diff
@@ -45,6 +45,8 @@ CHIP_ERROR SetValveLevel(chip::EndpointId ep, DataModel::Nullable<Percent> level
 CHIP_ERROR UpdateCurrentLevel(chip::EndpointId ep, chip::Percent currentLevel);
 CHIP_ERROR UpdateCurrentState(chip::EndpointId ep, ValveConfigurationAndControl::ValveStateEnum currentState);
 CHIP_ERROR EmitValveFault(chip::EndpointId ep, chip::BitMask<ValveConfigurationAndControl::ValveFaultBitmap> fault);
+CHIP_ERROR SetRemainingDurationExt(EndpointId endpoint, DataModel::Nullable<uint32_t> duration);
```
what are the `Ext` functions? For the others I understand they are `Get Attribute` however the attributes is `RemainingDuration` ... what does the `Ext` suffix mean?

andy31415 replied:
I see the CPP has the non ext functions, but not exposed. Why is that?

These things should be in doccomments if they make sense and in the PR summary as described changes.

sxb427 replied:
Thanks for the suggestion. I ended up exposing the pre-existing setRemainingDuration and getRemainingDuration, and deleted the duplicates.

andy31415 commented on examples/chef/common/clusters/valve-configuration-and-control/chef-valve-configuration-and-control-delegate-impl.cpp:
```diff
@@ -0,0 +1,165 @@
+            ChipLogError(DeviceLayer, "Unable to write RemainingDuration");
+            ret = chip::Protocols::InteractionModel::Status::Failure;
+        }
+        break;
```
remove this break or the other one outside of the {}

andy31415 commented on examples/chef/common/clusters/valve-configuration-and-control/chef-valve-configuration-and-control-delegate-impl.cpp:
```diff
@@ -0,0 +1,165 @@
+        }
+        if (duration.IsNull())
+        {
+            std::memset(buffer, '\0', maxReadLength);
```
is this correct? The null value for a uint32_t is via https://github.com/project-chip/connectedhomeip/blob/master/src/app/util/odd-sized-integers.h#L149 and generally 0xFFFFFFFF I believe.

andy31415 replied:
Please run a test to validate.

sxb427 replied:
Internally, the attribute is store as an optional, which allows NULL values. However, the API only accepts raw bytes. So inspired from [here](https://github.com/project-chip/connectedhomeip/blob/master/src/app/util/odd-sized-integers.h#L119), it seems 0xFF... interpreted as NULL is a known convention that the caller and callee can agree upon. So I've implemented it that way:
- If callee receives a write request of 0xFFF.., it simply sets the attr to NULL.
- If call receives a read request when the current attr. value is NULL, it simply returns 0xFFF..

andy31415 commented on examples/chef/common/clusters/valve-configuration-and-control/chef-valve-configuration-and-control-delegate-impl.cpp:
```diff
@@ -0,0 +1,165 @@
+        else
+        {
+            uint32_t val = duration.Value();
+            std::memcpy(buffer, &val, std::min(maxReadLength, static_cast<uint16_t>(sizeof(uint32_t))));
```
```suggestion
            std::memcpy(buffer, &val, std::min<size_t>(maxReadLength, sizeof(uint32_t)));
```

feels more readable somehow.

sxb427 replied:
Yes, Done.

andy31415 commented on examples/chef/common/clusters/valve-configuration-and-control/chef-valve-configuration-and-control-delegate-impl.cpp:
```diff
@@ -0,0 +1,165 @@
+                                             const EmberAfAttributeMetadata * attributeMetadata, uint8_t * buffer,
+                                             uint16_t maxReadLength)
+{
+    chip::Protocols::InteractionModel::Status ret = chip::Protocols::InteractionModel::Status::Success;
```
how about returning directly as needed instead of having a top variable that is used up to the end of the function? This style forces me to look from the bottom of the function through the entire implementation to follow things through and that is hard to follow. Use early returns as needed.

sxb427 replied:
Done.

andy31415 commented on src/app/clusters/valve-configuration-and-control-server/valve-configuration-and-control-server.cpp:
```diff
@@ -422,6 +422,25 @@ void UpdateAutoCloseTime(uint64_t time)
+
+CHIP_ERROR GetRemainingDurationExt(EndpointId endpoint, DataModel::Nullable<uint32_t> & duration)
+{
+    bool res = GetRemainingDuration(endpoint, duration);
```
why not make GetRemainingDuration return a CHIP_ERROR and expose it?

sxb427 replied:
Yeah that is a better idea. Those methods are exposed now. Duplicates (ending with Ext) deleted.

andy31415 commented on src/app/clusters/valve-configuration-and-control-server/valve-configuration-and-control-server.cpp:
```diff
@@ -422,6 +422,25 @@ void UpdateAutoCloseTime(uint64_t time)
+    SetRemainingDuration(endpoint, duration);
+    DataModel::Nullable<uint32_t> updatedDuration;
+    bool res = GetRemainingDuration(endpoint, updatedDuration);
+    if (!res || updatedDuration != duration)
```
when would `updateDuration` be not `duration`? Why this extra check?

How about making SetRemainingDuration return a CHIP_ERROR?

Should maybe caller then be responsible for the attribute change callback?

sxb427 replied:
Since SetRemainingDuration did not return an error code, this was the only way of determining success. Anyways, since it is okay to expose those, I've removed these duplicates.

sxb427 committed:
Merge branch 'project-chip:master' into valve-configuration-and-control

bzbarsky-apple commented on src/app/clusters/valve-configuration-and-control-server/valve-configuration-and-control-server.h:
```diff
@@ -45,6 +45,8 @@ CHIP_ERROR SetValveLevel(chip::EndpointId ep, DataModel::Nullable<Percent> level
 CHIP_ERROR UpdateCurrentLevel(chip::EndpointId ep, chip::Percent currentLevel);
 CHIP_ERROR UpdateCurrentState(chip::EndpointId ep, ValveConfigurationAndControl::ValveStateEnum currentState);
 CHIP_ERROR EmitValveFault(chip::EndpointId ep, chip::BitMask<ValveConfigurationAndControl::ValveFaultBitmap> fault);
+CHIP_ERROR SetRemainingDurationExt(EndpointId endpoint, DataModel::Nullable<uint32_t> duration);
```
Please name the API sanely.  If the implementation then needs changes because it's using the name internally, changing the internals is fine.

sxb427 replied:
No longer in use.

bzbarsky-apple commented on src/app/clusters/valve-configuration-and-control-server/valve-configuration-and-control-server.cpp:
```diff
@@ -422,6 +422,25 @@ void UpdateAutoCloseTime(uint64_t time)
+    DataModel::Nullable<uint32_t> updatedDuration;
+    bool res = GetRemainingDuration(endpoint, updatedDuration);
+    if (!res || updatedDuration != duration)
+        return CHIP_ERROR_WRITE_FAILED;
```
Shouldn't we just have the internal setter return a status or boolean, instead of doing this "set and then get" business?

sxb427 replied:
Yes, I've made the updates to do it the suggested way.

bzbarsky-apple commented on src/app/clusters/valve-configuration-and-control-server/valve-configuration-and-control-server.cpp:
```diff
@@ -422,6 +422,25 @@ void UpdateAutoCloseTime(uint64_t time)
+    bool res = GetRemainingDuration(endpoint, updatedDuration);
+    if (!res || updatedDuration != duration)
+        return CHIP_ERROR_WRITE_FAILED;
+    MatterReportingAttributeChangeCallback(endpoint, ValveConfigurationAndControl::Id, RemainingDuration::Id);
```
Why are we doing this even if the value has not changed?    I think this is because in practice it represents one of the Q conditions in the spec, maybe?  But this is very unclear and this part needs a bunch of documentation.  Including API documentation so callers know when they should/shouldn't call this setter.

sxb427 replied:
We could have added a check at the start to return if duration was equal to current duration. 
The only reason this API was created was to indirectly expose the pre-existing Set/GetRemainingTime methods, since I wasn't sure if exposing them was allowed.
Anyways, these duplicates have been removed and the current ones are now exposed by adding them inside ValveControlAndConfiguration namespace.

sxb427 committed:
* Expose GetRemainingDuration, SetRemainingDuration and SetRemainingDurationNull
* When reading remaining time, max numeric value for the given size (0xFFF...) is interpreted as NULL.

sxb427 committed:
Fix compilation errors

sxb427 committed:
Fix compile errors

sxb427 committed:
Fix compile errors

sxb427 committed:
Attribute change must be reported.

sxb427 committed:
Fix compile errors.

sxb427 committed:
Merge branch 'project-chip:master' into valve-configuration-and-control

sxb427 committed:
Merge branch 'master' into valve-configuration-and-control

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

andy31415 commented on examples/chef/common/clusters/valve-configuration-and-control/chef-valve-configuration-and-control-delegate-impl.cpp:
```diff
@@ -0,0 +1,168 @@
+        std::memcpy(&newVal, buffer, sizeof(uint32_t));
+        ChipLogProgress(DeviceLayer, "Setting RemainingDuration to %d", newVal);
+        CHIP_ERROR err;
+        if (newVal == static_cast<uint32_t>(0xFFFF)) // Max value is interpreted as NULL
```
Remaining duration is elapsed_s which is 32-bit (based on the memcpy above ... can we somehow enforce?)

So max value would be 0xFFFFFFFF ... for clarity maybe we should std::numeric_limits<uint32_t>::max() instead. The cast looks off (you could have appended a u)

sxb427 replied:
Changed to use `std::numeric_limits<uint32_t>::max()`.
I am not sure if there is a way to enforce the caller to pass in exactly 4 bytes. 

Larger than 4 bytes will truncate the data. But currently there is no 64-bit integer field in [AttributeData](https://github.com/project-chip/connectedhomeip/blob/master/examples/common/pigweed/protos/attributes_service.proto#L212-L225) filed so going over 4 bytes is not possible.

However if the buffer is smaller than 4 bytes (e.g. `data_int8`) then we may read garbage data. These handlers are used by PwRPC calls only and callers should abide by the sizes in matter spec.

andy31415 commented on examples/chef/common/clusters/valve-configuration-and-control/chef-valve-configuration-and-control-delegate-impl.cpp:
```diff
@@ -0,0 +1,168 @@
+        }
+        if (err != CHIP_NO_ERROR)
+        {
+            ChipLogError(DeviceLayer, "Unable to write RemainingDuration");
```
print a CHIP_ERROR_FORMAT of err as well please, otherwise debugging from logs will be hard.

sxb427 replied:
Done.

andy31415 commented on examples/chef/common/clusters/valve-configuration-and-control/chef-valve-configuration-and-control-delegate-impl.cpp:
```diff
@@ -0,0 +1,168 @@
+        CHIP_ERROR err = GetRemainingDuration(endpointId, duration);
+        if (err != CHIP_NO_ERROR)
+        {
+            ChipLogError(DeviceLayer, "Unable to read RemainingDuration");
```
also add the error to the log. we should do this throughout.

sxb427 replied:
Done.

andy31415 commented on examples/chef/common/clusters/valve-configuration-and-control/chef-valve-configuration-and-control-delegate-impl.cpp:
```diff
@@ -0,0 +1,168 @@
+        if (duration.IsNull())
+        {
+            // Max value is interpreted as NULL
+            std::memset(buffer, 0xFF, maxReadLength);
```
maybe use a std::min here as well?

andy31415 replied:
or just have a single memcpy and set val to the right value. like:

```cpp
uint32_t val = duration.ValueOr(std::numeric_limits<uint32_t>::max());
```

sxb427 replied:
Thanks for the suggestion. Looks cleaner now.

andy31415 commented on examples/chef/common/clusters/valve-configuration-and-control/chef-valve-configuration-and-control-delegate-impl.h:
```diff
@@ -0,0 +1,67 @@
+class DelegateImpl : public ValveConfigurationAndControl::Delegate
+{
+public:
+    DelegateImpl(){};
```
```suggestion
    DelegateImpl() = default;
```

otherwise no need for a `;` after `{}`. I would place constructor and destructor together.

sxb427 committed:
Minor changes to address comments

sxb427 committed:
Minor changes to address comments

sxb427 committed:
Fix compilation errors

sxb427 committed:
Merge branch 'master' into valve-configuration-and-control

sxb427 commented:
@bzbarsky-apple for the requested changes I exposed the pre-existing Get and Set methods and got rid of the duplicates. Can you take a look? Thanks!

andy31415 commented on src/app/clusters/valve-configuration-and-control-server/valve-configuration-and-control-server.h:
```diff
@@ -45,6 +45,9 @@ CHIP_ERROR SetValveLevel(chip::EndpointId ep, DataModel::Nullable<Percent> level
 CHIP_ERROR EmitValveFault(chip::EndpointId ep, chip::BitMask<ValveConfigurationAndControl::ValveFaultBitmap> fault);
+CHIP_ERROR GetRemainingDuration(EndpointId endpoint, DataModel::Nullable<uint32_t> & duration);
+CHIP_ERROR SetRemainingDuration(EndpointId endpoint, DataModel::Nullable<uint32_t> duration);
+CHIP_ERROR SetRemainingDurationNull(EndpointId endpoint);
```
Why do we have a `SetNull` when the other setter takes a nullable? What is the difference between calling `SetRemainingDurationNull` and `SetRemainingDuration(..., DataModel::NullOptional)`

andy31415 replied:
It seems pre-existing. Did not realize that. Valve seems quite broken :(

sxb427 replied:
Copied over form [pre-existing](https://github.com/project-chip/connectedhomeip/blob/master/src/app/clusters/valve-configuration-and-control-server/valve-configuration-and-control-server.cpp#L79-L102) code.

andy31415 commented on src/app/clusters/valve-configuration-and-control-server/valve-configuration-and-control-server.cpp:
```diff
@@ -422,6 +389,44 @@ void UpdateAutoCloseTime(uint64_t time)
+    {
+        if (!gRemainingDuration[epIdx].remainingDuration.IsNull())
+        {
+            MatterReportingAttributeChangeCallback(endpoint, ValveConfigurationAndControl::Id, RemainingDuration::Id);
```
Why is the attribute change callback done before setting the null value?

sxb427 replied:
Copied form [here](https://github.com/project-chip/connectedhomeip/blob/master/src/app/clusters/valve-configuration-and-control-server/valve-configuration-and-control-server.cpp#L90-L102).

andy31415 commented on src/app/clusters/valve-configuration-and-control-server/valve-configuration-and-control-server.cpp:
```diff
@@ -422,6 +389,44 @@ void UpdateAutoCloseTime(uint64_t time)
+    {
+        gRemainingDuration[epIdx].endpoint          = endpoint;
+        gRemainingDuration[epIdx].remainingDuration = duration;
+        return CHIP_NO_ERROR;
```
you are changing an attribute here. Why is MatterReportingAttributeChangeCallback not invoked here (but is on the set null version)?

andy31415 commented on src/app/clusters/valve-configuration-and-control-server/valve-configuration-and-control-server.cpp:
```diff
@@ -223,12 +190,12 @@ static void onValveConfigurationAndControlTick(System::Layer * systemLayer, void
     if (rDuration.Value() > 0)
     {
-        SetRemainingDuration(ep, DataModel::MakeNullable<uint32_t>(--rDuration.Value()));
+        ValveConfigurationAndControl::SetRemainingDuration(ep, DataModel::MakeNullable<uint32_t>(--rDuration.Value()));
```
why is the logic here using 0 for null when the value is a nullable already? Why is null not null?
Some comments explaining would be nice.

andy31415 commented on src/app/clusters/valve-configuration-and-control-server/valve-configuration-and-control-server.cpp:
```diff
@@ -145,7 +111,8 @@ ValveConfigAndControlAttrAccess gAttrAccess;
 {
     DataModel::Nullable<uint32_t> rDuration;
-    VerifyOrReturnError(GetRemainingDuration(endpoint, rDuration), CHIP_IM_GLOBAL_STATUS(UnsupportedAttribute));
+    VerifyOrReturnError(ValveConfigurationAndControl::GetRemainingDuration(endpoint, rDuration) == CHIP_NO_ERROR,
```
Why does every error result in UnsupportedAttribute? Should it not happen only for something like `NOT_IMPLEMENTED` or similar?

andy31415 commented on src/app/clusters/valve-configuration-and-control-server/valve-configuration-and-control-server.cpp:
```diff
@@ -422,6 +389,44 @@ void UpdateAutoCloseTime(uint64_t time)
+{
+    uint16_t epIdx = emberAfGetClusterServerEndpointIndex(endpoint, ValveConfigurationAndControl::Id,
+                                                          MATTER_DM_VALVE_CONFIGURATION_AND_CONTROL_CLUSTER_SERVER_ENDPOINT_COUNT);
+    VerifyOrReturnValue(epIdx < kValveConfigurationAndControlDelegateTableSize, CHIP_ERROR_READ_FAILED);
```
invalid argument or not found seem like better error codes. Maybe not found could be then mapped to "unsupported attribute".

You can also return some `CHIP_IM_GLOBAL_STATUS(UnsupportedAttribute)` directly here, so you do not do yet another re-mapping when using this.

andy31415 replied:
Pre-existing ... feel free to close

sxb427 replied:
I think `CHIP_ERROR_NOT_FOUND` is a better error code to return here.

andy31415 commented on examples/chef/common/clusters/valve-configuration-and-control/chef-valve-configuration-and-control-delegate-impl.cpp:
```diff
@@ -0,0 +1,161 @@
+    return CHIP_NO_ERROR;
+}
+
+void DelegateImpl::HandleRemainingDurationTick(uint32_t duration)
```
we need suffixes for sized values. Is this `duration_ms` or `duration_sec` or something else?

sxb427 replied:
Done/

andy31415 commented on examples/chef/common/clusters/valve-configuration-and-control/chef-valve-configuration-and-control-delegate-impl.cpp:
```diff
@@ -0,0 +1,161 @@
+
+    switch (attributeId)
+    {
+    case chip::app::Clusters::ValveConfigurationAndControl::Attributes::RemainingDuration::Id: {
```
```suggestion
    case Attributes::RemainingDuration::Id: {
```

You already use `chip::app::Clusters::ValveConfigurationAndControl`, so you can simplify the code throughout

sxb427 replied:
Done.

andy31415 commented on examples/chef/common/clusters/valve-configuration-and-control/chef-valve-configuration-and-control-delegate-impl.cpp:
```diff
@@ -0,0 +1,161 @@
+    {
+    case chip::app::Clusters::ValveConfigurationAndControl::Attributes::RemainingDuration::Id: {
+        uint32_t newVal = 0;
+        std::memcpy(&newVal, buffer, sizeof(uint32_t));
```
maybe using https://github.com/project-chip/connectedhomeip/blob/master/src/app/util/attribute-storage-null-handling.h#L42 is clearer. You will have `WorkingToStorage` and `StorageToWorking` as well as IsNullValue

sxb427 replied:
Done.

sxb427 committed:
Review changes:Use NumericAttributeTraits in Write callback.Fix nits.

sxb427 committed:
Cast uint8_t* buffer to uint32_t* buffer

sxb427 committed:
Fix compilation error: Use reinterpret_cast

sxb427 committed:
Fix formatting

bzbarsky-apple commented on src/app/clusters/valve-configuration-and-control-server/valve-configuration-and-control-server.h:
```diff
@@ -45,6 +45,9 @@ CHIP_ERROR SetValveLevel(chip::EndpointId ep, DataModel::Nullable<Percent> level
 CHIP_ERROR UpdateCurrentState(chip::EndpointId ep, ValveConfigurationAndControl::ValveStateEnum currentState);
 CHIP_ERROR EmitValveFault(chip::EndpointId ep, chip::BitMask<ValveConfigurationAndControl::ValveFaultBitmap> fault);
+CHIP_ERROR GetRemainingDuration(EndpointId endpoint, DataModel::Nullable<uint32_t> & duration);
+CHIP_ERROR SetRemainingDuration(EndpointId endpoint, DataModel::Nullable<uint32_t> duration);
```
```suggestion
CHIP_ERROR SetRemainingDuration(EndpointId endpoint, const DataModel::Nullable<uint32_t> & duration);
```

bzbarsky-apple commented on src/app/clusters/valve-configuration-and-control-server/valve-configuration-and-control-server.cpp:
```diff
@@ -422,6 +389,44 @@ void UpdateAutoCloseTime(uint64_t time)
+    if (epIdx < kValveConfigurationAndControlDelegateTableSize)
+    {
+        gRemainingDuration[epIdx].endpoint          = endpoint;
+        gRemainingDuration[epIdx].remainingDuration = duration;
```
Pre-existing, but isn't this missing the MatterReportingAttributeChangeCallback that should happen here?

Let's do a followup for this, because this should really be using QuieterReportingAttribute to get the right behavior.  Right now this stuff is just totally not spec-compliant... 

But that followup absolutely has to happen.

sxb427 replied:
I invoked `MatterReportingAttributeChangeCallback` at the caller side. But it makes more sense to call it here, if not  onValveConfigurationAndControlTick would not report a matter attr change.

I'm now moving the `MatterReportingAttributeChangeCallback` from caller to inside this function. Let me know what you think.

bzbarsky-apple commented on src/app/clusters/valve-configuration-and-control-server/valve-configuration-and-control-server.cpp:
```diff
@@ -422,6 +389,44 @@ void UpdateAutoCloseTime(uint64_t time)
+    {
+        if (!gRemainingDuration[epIdx].remainingDuration.IsNull())
+        {
+            MatterReportingAttributeChangeCallback(endpoint, ValveConfigurationAndControl::Id, RemainingDuration::Id);
```
Yeah, so this is a hack that implements _one_ of the 5 cases when this reporting needs to happen.  Looks to me like this code was either written before the spec firmed up or just randomly chose to not implement the spec for some reason....

Anyway: what this function, if it exists at all, should do is call `SetRemainingDuration(endpoint, NullNullable)` and not have any other logic.  This part _does_ need to happen in this PR (either removing the function or making it be a thin wrapper).  If that requires this hack to move into the 2-argument SetRemainingDuration, that should happen, I guess.

sxb427 replied:
I've removed logic in `SetRemainingDurationNull` and just call `SetRemainingDuration` with a null Nullable object.
Additionally added an early return in `SetRemainingDuration` to avoid calling `MatterReportingAttributeChangeCallback` if the new value is equal to the current value.

bzbarsky-apple commented on examples/chef/common/clusters/valve-configuration-and-control/chef-valve-configuration-and-control-delegate-impl.cpp:
```diff
@@ -0,0 +1,160 @@
+
+static DelegateImpl * gValveConfigurationAndControlDelegate = nullptr;
+
+DataModel::Nullable<chip::Percent> DelegateImpl::HandleOpenValve(DataModel::Nullable<chip::Percent> level)
```
Strongly recommend making the argument a const ref, but won't block on that.

bzbarsky-apple commented on examples/chef/common/clusters/valve-configuration-and-control/chef-valve-configuration-and-control-delegate-impl.cpp:
```diff
@@ -0,0 +1,160 @@
+    case Attributes::RemainingDuration::Id: {
+        CHIP_ERROR err;
+        uint32_t * bufUint32 = reinterpret_cast<uint32_t *>(buffer);
+        if (NumericAttributeTraits<uint32_t>::IsNullValue(*bufUint32)) // Max value is interpreted as NULL
```
```suggestion
        if (NumericAttributeTraits<uint32_t>::IsNullValue(*bufUint32))
```
Whatever the null value is is interpreted as null.  From this code's point of view, the fact that it's the "max value" is an irrelevant implementation detail.

sxb427 replied:
Done.

bzbarsky-apple commented on examples/chef/common/clusters/valve-configuration-and-control/chef-valve-configuration-and-control-delegate-impl.cpp:
```diff
@@ -0,0 +1,160 @@
+    case Attributes::RemainingDuration::Id: {
+        CHIP_ERROR err;
+        uint32_t * bufUint32 = reinterpret_cast<uint32_t *>(buffer);
+        if (NumericAttributeTraits<uint32_t>::IsNullValue(*bufUint32)) // Max value is interpreted as NULL
```
This can lead to unaligned access and crashes, no?  I think you need to memcpy the data into an on-stack uint32_t here.

sxb427 replied:
Yes doing a `*bufUint32` can lead to restricted memory access if the allocated buffer size is less than 4 bytes. Here are the reasons for doing it this way -
* PwRPC clients like RPC-console need to explicitly specify attribute type when calling read/write (INT32U in this case). So theres low chance of wrong data size buffer.
* This function will probably be deprecated after we implement a `Write` method on the `ValveConfigAndControlAttrAccess` class. `DataModel::Provider` tries to use AAI before ember callbacks.
* I'd trust available libraries to take care of endianness.

bzbarsky-apple commented on examples/chef/common/clusters/valve-configuration-and-control/chef-valve-configuration-and-control-delegate-impl.cpp:
```diff
@@ -0,0 +1,160 @@
+        if (NumericAttributeTraits<uint32_t>::IsNullValue(*bufUint32)) // Max value is interpreted as NULL
+        {
+            ChipLogProgress(DeviceLayer, "Setting RemainingDuration to NULL.");
+            err = SetRemainingDurationNull(endpointId);
```
This is ... really broken.  Why isn't the cluster handling this?  This shouldn't need to be handled by the delegate. 

And in fact, the cluster does not actually use the attribute store for this stuff at all, so when would this code even get invoked?

sxb427 replied:
Currently there is no `Write` implemented on `ValveConfigAndControlAttrAccess` class thats why we need this callback (`DataModel::Provider` will end up calling this if it didn't find the corresponding call on AAI).

The right way to go about this is to implement `Write` on `ValveConfigAndControlAttrAccess` and get rid of this function. However..., at the time I wrote this function, the `Write` function in [rpc_services/Attributes.h](https://github.com/project-chip/connectedhomeip/blob/master/examples/common/pigweed/rpc_services/Attributes.h) directly called ember instead of doing it through `DataModel::Provider`. As a result AAI calls would be unused when write calls were made by RPC server. And thats why this ember function needed to be defined. This defect was fixed in PR: [37610](https://github.com/project-chip/connectedhomeip/pull/37160).

I plan to remove this callback in a future PR. Ideally chef delegates shouldn't have to write ember callbacks at all. All externally managed attributes should be accessed through AAI, and AAI Read/Writes should be a part of the cluster code.
Since `rpc_services` didn't use `DataModel::Provider` before, we had to write the ember callbacks. Anyways now `rpc_services` deficiencies have been fixed by @andreilitvin .

bzbarsky-apple commented on examples/chef/common/clusters/valve-configuration-and-control/chef-valve-configuration-and-control-delegate-impl.cpp:
```diff
@@ -0,0 +1,160 @@
+            ChipLogError(DeviceLayer, "Unable to write RemainingDuration: %" CHIP_ERROR_FORMAT, err.Format());
+            return chip::Protocols::InteractionModel::Status::Failure;
+        }
+        MatterReportingAttributeChangeCallback(endpointId, ValveConfigurationAndControl::Id, Attributes::RemainingDuration::Id);
```
This right here is a spec violation, and a serious one.  The good news is that this code should not exist, so that will solve that problem.

sxb427 replied:
I've moved the `MatterReportingAttributeChangeCallback` over to the server code. And the rest is explained in [comment](https://github.com/project-chip/connectedhomeip/pull/37088/files/7562eb0dee2979d20941e78c43e6f10c30f530a4#r1931309859).

bzbarsky-apple commented on examples/chef/common/clusters/valve-configuration-and-control/chef-valve-configuration-and-control-delegate-impl.cpp:
```diff
@@ -0,0 +1,160 @@
+
+    switch (attributeId)
+    {
+    case Attributes::RemainingDuration::Id: {
```
Was this tested?  As far as I can tell, this is dead code that will never be reached, since the AttributeAccessInterface on the cluster will always handle this.

sxb427 replied:
Yes I tested this and it was never called. Even RPC has migrated to use `DataModel::Provider`, so RPC clients also don't need this anymore.

bzbarsky-apple commented on examples/chef/common/clusters/valve-configuration-and-control/chef-valve-configuration-and-control-delegate-impl.cpp:
```diff
@@ -0,0 +1,160 @@
+void emberAfValveConfigurationAndControlClusterInitCallback(chip::EndpointId endpointId)
+{
+    // Currently this cluster delegate handles only one endpoint.
+    VerifyOrDieWithMsg(endpointId == 1, DeviceLayer, "ValveConfigurationAndControl cluster is only eabled for endpoint 1.");
```
```suggestion
    VerifyOrDieWithMsg(endpointId == 1, DeviceLayer, "ValveConfigurationAndControl cluster is only enabled for endpoint 1.");
```

sxb427 committed:
Merge branch 'project-chip:master' into valve-configuration-and-control

sxb427 committed:
Update src/app/clusters/valve-configuration-and-control-server/valve-configuration-and-control-server.h

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

sxb427 committed:
Invoke matter change attribute callback in setter

sxb427 committed:
fix compilation errors

sxb427 committed:
Remove un-necessary comment detail

sxb427 committed:
Update examples/chef/common/clusters/valve-configuration-and-control/chef-valve-configuration-and-control-delegate-impl.cpp

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

sxb427 committed:
Use SetRemainingDuration inside SetRemainingDurationNull

sxb427 commented on src/app/clusters/valve-configuration-and-control-server/valve-configuration-and-control-server.h:
```diff
@@ -45,6 +45,9 @@ CHIP_ERROR SetValveLevel(chip::EndpointId ep, DataModel::Nullable<Percent> level
 CHIP_ERROR EmitValveFault(chip::EndpointId ep, chip::BitMask<ValveConfigurationAndControl::ValveFaultBitmap> fault);
+CHIP_ERROR GetRemainingDuration(EndpointId endpoint, DataModel::Nullable<uint32_t> & duration);
+CHIP_ERROR SetRemainingDuration(EndpointId endpoint, DataModel::Nullable<uint32_t> duration);
+CHIP_ERROR SetRemainingDurationNull(EndpointId endpoint);
```
I needed to access `RemainingDurationTable gRemainingDuration[]` from external delegate code. Thats why the APIs had to be exposed.

sxb427 committed:
Implement Write on AAI

sxb427 commented:
@bzbarsky-apple below is the summary of changes since the last comment set -
1. Moved `MatterReportingAttributeChangeCallback` to `SetRemainingDuration`.
2. Replaced logic in `SetRemainingDurationNull` with call to `SetRemainingDuration(NullNullable)`.
3. Implemented `Write` method on `ValveConfigAndControlAttrAccess`. Verified all Reads and Writes both via chip tool and RPC go through AAI first. Now both the ember read/write methods are never called, but I'd still like to keep them for now until we've decided to fully rely on AAI.

sxb427 committed:
Merge branch 'project-chip:master' into valve-configuration-and-control

tcarmelveilleux reviewed (changes requested):
[ ] Do no implement write for attributes that are not writable

sxb427 committed:
Remove WriteRemainingDuration as it violates spec by allowing writes to a read-only atribute

sxb427 committed:
Merge branch 'project-chip:master' into valve-configuration-and-control

sxb427 committed:
Merge branch 'project-chip:master' into valve-configuration-and-control
